### PR TITLE
Closes a few issues

### PR DESCRIPTION
**Pull request by r-salas**
Add more info to log, add platform details, reorganize var and preprocessors directories, merge preprocessors and transformers, fix MTE API, remove -c argument when executing from docker, option to read JSON for Kaggle credentials

### CODE REVIEW

**Review by kawaits — CHANGES_REQUESTED** [review]
Please update **documentation**:
 - session now is part of the model config:
        - architecture.rst#training
        - architecture.rst#session
        - architecture.rst#system configuration
        - examples/howto-model-handler.rst
 - fleximodel:
        - architecture.rst#models 
        - examples/howto-model-handler.rst
        - samples-> from training to dataset
           - explain implications for validation samples (no longer handled by the trainer)
        - explain new param "seed" and "shuffle" options
        - explain pros and cons of "duck-typing"
        - document new utils
        - document input_shape now part of model specific parameters
 - trackers:
        - complete current info at tracking.rst
 - platform, system and HW metrics CPU/GPU:
        - extend execution-results.rst metrics section
 - keras metrics:
        - extend execution-results.rst metrics section
 - preprocessors:
        - extend architecture.rst#datasets
 - kaggle datasets:
        - original documentation included info about 
           - how to proceed with kaggle datasets in terms of credentials
           - which datasets were supported
        - please recover this info and extend  datasets.rst with any additional recommendation (ie: when data loading fails due to a credential error )
 - nvidia-docker:
        - extend installing-gymnos.rst with info about how to get nvidia-docker, install it and how to get it up and running

File: src/experiments/mnist.json [hunk 1]
```diff
@@ -5,21 +5,28 @@
...
-            "metrics": ["acc"]
-        },
-        "network": [
-            {"type": "flatten"},
-            {"type": "dense", "units": 1024, "activation": "relu"},
-            {"type": "dense", "units": 512, "activation": "relu"},
-            {"type": "dense", "units": 64, "activation": "relu"},
-            {"type": "dense", "units": 10, "activation": "softmax"}
-        ]
+        "name": "keras",
```
Comment (kawaits) [inline src/experiments/mnist.json:8]: 1.- Experiment should be agnostic to comercial libraries.
2.- Name should be descriptive of the model not the library.  name:" keras" seems not to provide much info
Please find alternative. We can discuss it later

File: src/experiments/tiny_imagenet.json [hunk 2]
```diff
@@ -5,23 +5,30 @@
...
-            {"type": "max_pooling_2d", "pool_size": 2},
-            {"type": "flatten"},
-            {"type": "dense", "units": 512, "activation": "relu"},
-            {"type": "dense", "units": 200, "activation": "softmax"}
-        ],
-        "compilation": {
-            "optimizer": "adam",
-            "loss": "categorical_crossentropy",
-            "metrics": ["acc"]
+        "name": "keras",
```
Comment (kawaits) [inline src/experiments/tiny_imagenet.json]: Idem

File: src/experiments/unusual_data_usage_test.json [hunk 3]
```diff
@@ -11,16 +11,10 @@
...
-    "session": {
-        "device": "cpu"
-    },
-    "training": {
+        "name": "unusual_data_usage_test",
         "samples": {
-            "fit": 0.20,
-            "val": 0.20,
-            "test": 0.20
+            "train": 0.8,
```
Comment (kawaits) [inline src/experiments/unusual_data_usage_test.json]: I do not understand this: 
I thought from now on, samples were part of the "dataset" section not "training"

File: src/lib/models/keras.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+class Keras(Model, KerasMixin):
+
+    def __init__(self, sequential, compilation):
+        loss, optimizer, metrics = self.__build_compilation_from_config(**compilation)
+        self.model = self.__build_sequential_from_config(sequential)
+        self.model.compile(loss=loss, optimizer=optimizer, metrics=metrics)
+
+
+    def __build_compilation_from_config(self, loss, optimizer, metrics=None):
```
Comment (kawaits) [inline src/lib/models/keras.py:46]: Why this method is not part of KerasMixin. What is the criteria?

File: src/lib/models/keras.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            if metric_name in metrics_ids_to_modules:
+                metric_func = locate(metrics_ids_to_modules[metric_name])
+                metrics_funcs.append(metric_func)
+            else:
+                metrics_funcs.append(metric_name)
+
+        return loss, optimizer, metrics_funcs
+
+
+    def __build_sequential_from_config(self, sequential_config):
```
Comment (kawaits) [inline src/lib/models/keras.py:66]: Why this method is not part of KerasMixin. What is the criteria?

File: src/lib/models/model.py [hunk 6]
```diff
@@ -4,86 +4,21 @@
...
+class Model(BaseEstimator):
+    """
+    Base model for all gymnos models, you must implement the following methods:
+        - fit(**parameters) -> dict
+        - predict(X) -> array
+        - evaluate(X, y) -> dict
+        - save(directory)
+        - restore(directory)
+
+    We use duck-typing but this class is defined with 2 purposes:
```
Comment (kawaits) [inline src/lib/models/model.py]: What's the advantage of having duck typing?
Duck typing is **more about convenience than type safety**. Source: Cui, 2015. 
It's been said that dynamic typing in general cuts down development time. There's less boilerplate code. Code is easier to hack. With static typing, compile-time checks slow down the development process. Static typing could be used at a later point to get better performance. **Others recommend duck typing only for prototyping and never for production.**

While it's true that compile-time checks are useful to catch potential problems early, duck typing enforces following coding conventions, documentation and test-driven methodologies.

Since duck typing isn't exactly a type system, it gives programmers flexibility. In Python, for example, common things are simpler to code. While static typed languages use interfaces, such interfaces may involve excessive refactoring of client code.

File: src/lib/trainer.py [hunk 7]
```diff
@@ -36,16 +38,17 @@ class Trainer:
...
+        │       │   └── training.callbacks[i].name/
+        │       ├── TRAINING_METRICS_FILENAME  # filename to keep metrics generated and elapsed times
+        │       ├── TRAINING_MODEL_DIRECTORY  # model weights/parameters
+        │       │   └── ... # artifacts saved by model
         └── TRAINING_TRACKINGS_FOLDERNAME/  # artifacts generated by each tracker
             └── tracking.trackers[i].name/
     """
 
-    def __init__(self, experiment, model, dataset, training, session=None, tracking=None):
+    def __init__(self, experiment, model, dataset, training, session, tracking):
```
Comment (kawaits) [inline src/lib/trainer.py]: Why we are still having session as an input for the trainer? It is used no where. 
No a model class responsability
Please modify accordingly

File: src/bin/scripts/gymnosd.py [hunk 8]
```diff
@@ -45,10 +45,10 @@
     os.makedirs(cache_config["datasets"], exist_ok=True)
 
     trainer = Trainer(
-        experiment=Experiment(**training_config["experiment"]),
         model=Model(**training_config["model"]),
         dataset=Dataset(cache_dir=cache_config["datasets"], **training_config["dataset"]),
         training=Training(**training_config["training"]),
+        experiment=Experiment(**training_config.get("experiment", {})),  # optional field
```
Comment (kawaits) [inline src/bin/scripts/gymnosd.py]: In which scenario experiment would not be required?

Comment (r-salas) [reply src/bin/scripts/gymnosd.py]: Right now, name and tags for experiment field are optional fields, so we could do something like this:
```json
{
   "experiment": {}
}
```
I don't think is a good idea to force to have this field if both attributes are optional.

File: src/experiments/unusual_data_usage_test.json (hunk 3 above)
Comment (r-salas) [reply src/experiments/unusual_data_usage_test.json]: That's how it is.

File: src/lib/models/keras.py (hunk 4 above)
Comment (r-salas) [reply src/lib/models/keras.py:46]: You're right. I'm getting on it.

File: src/lib/models/keras.py (hunk 5 above)
Comment (r-salas) [reply src/lib/models/keras.py:66]: You're right. I'm getting on it.

File: src/lib/models/keras.py (hunk 4 above)
Comment (r-salas) [reply src/lib/models/keras.py:46]: I'm reviewing the files and they are correctly defined. The keras mixin expects a `self.model` variable with a compiled keras model, it does not have a constructor. The keras model is the responsible for defining the model from the sequential and compilation parameters.

File: src/lib/models/keras.py (hunk 5 above)
Comment (r-salas) [reply src/lib/models/keras.py:66]: Idem

File: Pipfile.lock [hunk 9]
```diff
@@ -43,17 +43,17 @@
         },
         "boto3": {
```
Comment (kawaits) [inline Pipfile.lock:44]: ¿AWS?

File: Pipfile.lock [hunk 10]
```diff
@@ -43,17 +43,17 @@
...
             "hashes": [
-                "sha256:c2816b25fdd792d759ef4d19b6dd551ef93232992d55fbf87307a71ab62ee159",
-                "sha256:f6f58cf19cc99bc90f1c57dcddf6fdf97d63e152263fc32bfa1d45ee4b278009"
+                "sha256:883f7143bcb081a834f7c09659524059b66745ea043fffd40420e88ef0143feb",
+                "sha256:9c789a775f0499743b083ffd63e0e87dae9a727511bb37f2529da52ccd25a360"
             ],
-            "version": "==1.9.131"
+            "version": "==1.9.134"
         },
         "botocore": {
```
Comment (kawaits) [inline Pipfile.lock:51]: ¿AWS?

File: docs/source/trackers.rst [hunk 11]
```diff
@@ -0,0 +1,22 @@
+###############################
```
Comment (kawaits) [inline docs/source/trackers.rst:1]: Please combine with  https://github.com/Telefonica/gymnos/blob/devel/docs/source/tracking.rst

File: Pipfile.lock (hunk 9 above)
Comment (r-salas) [reply Pipfile.lock:44]: MLFlow dependency.

File: Pipfile.lock (hunk 10 above)
Comment (r-salas) [reply Pipfile.lock:51]: MLFlow dependency

**Comment by r-salas** [comment]
This branch can be merged. 
New features:
- Regression test (#46)
- Updated and extended documentation (#16, #55, #66)
- HW details for metrics and logging (#64)
- Tensorflow session builder (#64)
- Preprocessors short logging (#67)
- Metrics short logging (#68)
- Remove Docker -c option (#69)
- Model flexibility (#73)
- Dataset flexibility
- Fix M+ API MTE (#74)
- Better handling Kaggle credentials (#75)
- Unify preprocessors and transformers (#81)
- Closes #82 (outdated)
- Updated TensorFlow version (#88)
- Add data usage model to regression test (#54)
